src/main/java/net/logstash/logback/pattern/AbstractJsonPatternParser.java

```diff
@@ -85,76 +87,59 @@ public AbstractJsonPatternParser(final ContextAware contextAware, final JsonFact
      * @param name the name of the operation
      * @param operation the {@link Operation} instance
      */
-    protected void addOperation(String name, Operation<?> operation) {
+    protected void addOperation(String name, Operation<Event, ?> operation) {
         this.operations.put(name, operation);
     }
 
-    protected abstract class Operation<T> {
-        private final boolean requiresData;
-
-        Operation(boolean requiresData) {
-            this.requiresData = requiresData;
-        }
-
-        public boolean requiresData() {
-            return requiresData;
-        }
-
-        public abstract ValueGetter<T, Event> createValueGetter(String data);
+    protected interface Operation<Event, T> {
+        /**
+         * Create a {@link ValueGetter} implementing the operation on the supplied arguments ({@code data}).
+         * Throws an {@link IllegalArgumentException} when the supplied data is invalid.
+         * 
+         * @param data the data to pass to the operation
+         * @return a {@link ValueGetter} implementing the operation
+         * @throws IllegalArgumentException if the supplied data is invalid
+         */
+        ValueGetter<Event, T> createValueGetter(String data);
     }
 
-    protected class AsLongOperation extends Operation<Long> {
-        public AsLongOperation() {
-            super(true);
-        }
-
+    protected abstract class TransformingOperation<T> implements Operation<Event, T> {
         @Override
-        public ValueGetter<Long, Event> createValueGetter(String data) {
-            return makeLayoutValueGetter(data).andThen(Long::parseLong);
+        public ValueGetter<Event, T> createValueGetter(String data) {
+            return makeLayoutValueGetter(data).andThen(this::convert);
         }
+        
+        protected abstract T convert(String value);
```

Comment:
_nit_ perhaps rename to `transform` so the method name relates to the class name